### PR TITLE
planner: handle PointPlan specially in execute statements

### LINKED ISSUE #36783: Performance regression of BenchmarkPreparedPointGet
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

See http://www.zenlife.tk:18081/

image

### 2. What did you expect to see? (Required)

No performance regression

### 3. What did you see instead (Required)

Performance regression on BenchmarkPreparedPointGet

### 4. What is your TiDB version? (Required)

master 9a16a8dd19b7daa5b0c418b5772456973b5001f5

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: close #36783

Problem Summary: planner: handle PointPlan specially in execute statements

### What is changed and how it works?

Following https://github.com/pingcap/tidb/pull/36587, handle PointPlan specially in execute statements to gain some performance improvement.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/plan_cache.go [hunk 1]
```diff
@@ -631,3 +632,42 @@ func GetBindSQL4PlanCache(sctx sessionctx.Context, preparedStmt *CachedPrepareSt
 	}
 	return "", ignore
 }
+
+// IsPointPlanShortPathOK check if we can execute using plan cached in prepared structure
+// Be careful with the short path, current precondition is ths cached plan satisfying
+// IsPointGetWithPKOrUniqueKeyByAutoCommit
+func IsPointPlanShortPathOK(sctx sessionctx.Context, preparedStmt *CachedPrepareStmt) (bool, error) {
```
Comment (qw4990) [inline planner/core/plan_cache.go:639]: From https://github.com/pingcap/tidb/pull/36587/files#diff-10dab83c04a99358d431906d37f48b92be97d50a2719d4ddf81c62d3b03a0aebL2374

File: session/session.go [hunk 2]
```diff
@@ -2302,6 +2302,25 @@ func (s *session) preparedStmtExec(ctx context.Context, execStmt *ast.ExecuteStm
 	}
 	sessionExecuteCompileDurationGeneral.Observe(time.Since(s.sessionVars.StartTime).Seconds())
 	logGeneralQuery(st, s, true)
+
+	pointShortPathOK, err := plannercore.IsPointPlanShortPathOK(s, prepareStmt)
+	if err != nil {
+		return nil, err
+	}
+	if pointShortPathOK {
```
Comment (qw4990) [inline session/session.go:2310]: From https://github.com/pingcap/tidb/pull/36587/files#diff-10dab83c04a99358d431906d37f48b92be97d50a2719d4ddf81c62d3b03a0aebL2353

Comment (xhebox) [reply session/session.go:2310]: I remembered that `case nil` is added intentionly, https://github.com/Reminiscent/tidb/commit/92b1b8e0e9598952917ab16fff0e17bcdac74992
Safe to remove?

**Review by lcwangchao — APPROVED** [review]

**Comment by qw4990** [comment]
Master v.s. this PR:
```
MASTER
BenchmarkPreparedPointGet-8       143991              8376 ns/op
THIS PR
BenchmarkPreparedPointGet-8       199149              5908 ns/op
```

File: executor/prepared.go [hunk 3]
```diff
@@ -346,6 +346,23 @@ func CompileExecutePreparedStmt(ctx context.Context, sctx sessionctx.Context,
...
+			if pointPlan, ok := ep.Plan.(*plannercore.PointGetPlan); ok {
+				// set for point plan short path
+				stmtCtx := sctx.GetSessionVars().StmtCtx
+				stmt.Text = preparedObj.PreparedAst.Stmt.Text()
+				stmtCtx.OriginalSQL = stmt.Text
+				stmtCtx.SetPlan(execPlan)
+				stmtCtx.InitSQLDigest(preparedObj.NormalizedSQL, preparedObj.SQLDigest)
+				stmtCtx.SetPlanDigest(preparedObj.NormalizedPlan, preparedObj.PlanDigest)
+				stmt.Plan = pointPlan
+				stmt.PsStmt = preparedObj
```
Comment (Reminiscent) [inline executor/prepared.go:361]: Should we return here?

Comment (qw4990) [reply executor/prepared.go:362]: udpated, ptal

**Review by Reminiscent — APPROVED** [review]

File: session/session.go (hunk 2 above)
Comment (qw4990) [reply session/session.go]: We can ignore this case-branch since `runStmt` is used by default.

**Comment by hawkingrei** [comment]
/hold 

Please fix the unit test.

**Pull request merged**